### PR TITLE
Config refactor

### PR DESCRIPTION
**Pull request by gfgtdf**

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
Merge soon?

File: src/config.cpp [hunk 1]
```diff
@@ -1042,49 +1037,7 @@ namespace {
 void config::clear()
```
Comment (CelticMinstrel) [inline src/config.cpp]: Please remove the `struct config_clear_state`.

File: src/config.cpp [hunk 2]
```diff
@@ -511,7 +511,7 @@ void config::recursive_clear_value(config_key_type key)
 	map_erase_key(values_, key);
 
 	for(std::pair<const std::string, child_list>& p : children_) {
-		for(config* cfg : p.second) {
+		for(auto& cfg : p.second) {
```
Comment (CelticMinstrel) [inline src/config.cpp]: No real reason for `auto&` over `auto` here to be honest.

Comment (gfgtdf) [inline src/config.cpp]: well sure, auto would copy the object which fails to complie since unique_ptr has no cctor.

Comment (CelticMinstrel) [inline src/config.cpp]: Oh, it was changed to `unique_ptr`. Okay then.

**Pull request merged**